trie/sync_test.go

```diff
@@ -585,16 +588,32 @@ func testIncompleteSync(t *testing.T, scheme string) {
 			})
 		}
 	}
+
+	// Cache trie nodes for faster verification
+	nodesCache := make(map[string][]byte)
+	resolver := func(owner common.Hash, path []byte, hash common.Hash) []byte {
+		if scheme == rawdb.HashScheme {
+			return nodesCache[string(hash[:])]
+		}
+		return nodesCache[string(path)]
+	}
+
 	// Sanity check that removing any node from the database is detected
 	for i, path := range addedKeys {
 		owner, inner := ResolvePath([]byte(path))
 		nodeHash := addedHashes[i]
 		value := rawdb.ReadTrieNode(diskdb, owner, inner, nodeHash, scheme)
 		rawdb.DeleteTrieNode(diskdb, owner, inner, nodeHash, scheme)
-		if err := checkTrieConsistency(diskdb, srcDb.Scheme(), root, false); err == nil {
+		if err := checkTrieConsistency(diskdb, srcDb.Scheme(), root, false, resolver); err == nil {
```

Comment:
Interesting! This _does_ indeed seriously improve the time, but it's pretty strange, IMO. I mean, the 'disk' is actually a memory database, so it's a `string->[]byte` mapping which you are basically replacing with a different `string->[]byte` mapping.

Reply:
Yes, but one of them needs to lookup stuff in the trie, meaning a read amplification because you need to go down the tree in your mapping, so O(log(n)) instead of O(1)

Reply:
I think adding a resolver here, while yes it apparently does make a difference, just makes the code more complicated and makes it harder to reason about what the test is testing. 

Now, what this part does, is that it goes through ~4000  (nodes? leaves?) and, one by one, 
- deletes it from the db, 
- iterates the entire trie until it hits an error
- adds it back to the db. 

And it does this both for hash and path-based. 

This test is a bit overzealous, IMO. Maybe we could  just pick 100 random elements and do this on? I mean, we check that "yes, indeed, removing an element makes iterating the trie impossible". It would be strange indeed if that check failed, so while a quick sanity-check is good, I don't see the need to exhaustively _prove_ it.